src/adapters/thinkerbell/mod.rs

```diff
@@ -0,0 +1,404 @@
+//! An adapter providing access to the Thinkerbell rules engine.
+
+use foxbox_taxonomy::adapter::{ AdapterManagerHandle, Adapter, WatchEvent, AdapterWatchGuard };
+use foxbox_taxonomy::api::{ API, Error, InternalError, ResultMap };
+use foxbox_taxonomy::services::{ Setter, Getter, AdapterId, ServiceId, Service, Channel, ChannelKind };
+use foxbox_taxonomy::util::Id;
+use foxbox_taxonomy::values::{ Range, Duration, Type, Value, TypeError, OnOff };
+use foxbox_thinkerbell::compile::ExecutableDevEnv;
+use foxbox_thinkerbell::manager::{ ScriptManager, ScriptId, Error as ScriptManagerError };
+use foxbox_thinkerbell::run::ExecutionEvent;
+
+use timer;
+use transformable_channels::mpsc::*;
+
+use std::collections::{ HashMap, HashSet };
+use std::path::Path;
+use std::sync::{ Arc, Mutex };
+use std::thread;
+
+static ADAPTER_NAME: &'static str = "Thinkerbell adapter (built-in)";
+static ADAPTER_VENDOR: &'static str = "team@link.mozilla.org";
+static ADAPTER_VERSION: [u32;4] = [0, 0, 0, 0];
+
+/// ThinkerbellAdapter hooks up the rules engine (if this, then that) as an adapter.
+///
+/// Each "rule", or "script", is a JSON-serialized structure according to Thinkerbell conventions.
+///
+/// This adapter exposes a root service, with one AddThinkerbellRule setter (to add a new rule).
+/// Each rule that has been added is exposed as its own service, with the following getters/setters:
+/// - Set Enabled (setter) -- toggles whether or not the script is enabled
+/// - Get Enabled (getter) -- returns whether or not the script is enabled
+/// - Remove (setter) -- removes the script
+///
+/// This adapter performs most actions by delegating channel messages to its main thread.
+#[derive(Clone)]
+pub struct ThinkerbellAdapter<T> where T: API + Clone + Send + 'static {
+    /// The sending end of the channel for sending messages to ThinkerbellAdapter's main loop.
+    tx: RawSender<ThinkAction>,
+    /// A reference to the AdapterManager.
+    adapter_manager: T,
+    /// The ID of this adapter (permanently fixed)
+    adapter_id: Id<AdapterId>,
+    /// The ID of the root service's "Add Rule" setter.
+    setter_add_rule_id: Id<Setter>,
+}
+
+/// Thinkerbell requires an execution environment following this API.
+#[derive(Clone)]
+struct ThinkerbellExecutionEnv<T> {
+    adapter_manager: T,
+    // FIXME: Timer's not clonable, so we should only use one, right? Does this have to be mutexed?
```

Comment:
Unless `ThinkerbellExecutionEnv` needs to be `Sync` (and I'm almost sure that's not the case), you don't need to `Mutex` it.